flash_attn/cute/flash_fwd.py

```diff
@@ -580,27 +589,33 @@ def __call__(
             cute.size(mQ.shape[2]),
             cute.size(mQ.shape[3]),
         )
-        # If there's tanh softcapping, we do tanh(scores * softmax_scale / softcap_val) * softcap_val.
-        # Right after this, we multiply by log2(e) before applying exp2.
-        # To reduce the number of instructions, we instead pre-multiply softmax_scale / softcap_val
-        # (assigning it to softcap_val) and pre-multiply softcap_val * log2(e)
-        # (assigning it to softmax_scale_log2).
         LOG2_E = math.log2(math.e)
-        if const_expr(softcap is None):
-            softmax_scale_log2 = softmax_scale * LOG2_E
-            softcap_val = None
+        if const_expr(self.score_mod is None):
+            softmax_scale_log2 = Float32(softmax_scale * LOG2_E)
+            softmax_scale = None
         else:
-            softmax_scale_log2 = softcap * LOG2_E
-            softcap_val = Float32(softmax_scale / softcap)
-            
+            # NB: If a user passes in a score mod, we want to apply the score-mod in the sm_scaled qk
+            # But in the original base 10. We hijack softmax_scale_log2 to just be the change of base
+            # and correctly apply the softmax_scale prior to score_mod in the softmax step
+            softmax_scale_log2 = Float32(LOG2_E)
+            softmax_scale = Float32(softmax_scale)
+
+        fastdiv_mods = None
+        if cutlass.const_expr(buffers is not None):
+            seqlen_q = cute.size(mQ.shape[0])
```

Comment:
This wouldn't work for varlen case right?
Do we focus on the non-varlen case for now?

Reply:
Maybe for the varlen case we set the divmod to be None, and inside the kernel do `if const_expr`. But if we don't care about the varlen case for now we can just leave this as is.